slotsmanager.c

```diff
@@ -141,32 +141,41 @@ slots_gc(void)
 static int
 slots_gc_base()
 {
-    int i, tree_invalid = 0; 
-    int count_elim;
     int64_t total0, total;
 
     sumup_large_ints(1, &NumPart, &total0);
 
-    count_elim = 0;
-
-    for(i = 0; i < NumPart; i++)
-        if(P[i].IsGarbage)
-        {
-            P[i] = P[NumPart - 1];
-
-            NumPart--;
-            i--;
-
-            count_elim++;
+    if(SlotsManager->garbage) {
+        int i, pc = 0;
+        int * partBufcnt = mymalloc("parttmp", (SlotsManager->garbage+1) * sizeof(int));
+        for(i = 0; i < NumPart; i++)
+            if(P[i].IsGarbage) {
+                /* mark the particle for removal in base slot.*/
+                partBufcnt[pc++] = i;
+            }
+
+        /*Set final elements*/
+        partBufcnt[pc] = NumPart;
+
+        /*Compactify the P array: this invalidates the ReverseLink.*/
```

Comment:
.. invalidates the ReverseLink causing ReverseLink only valid within slots_gc_slots.